### PR TITLE
fix: support nested slots with wrappers

### LINKED ISSUE #48: Support slotted inputs properly
Currently `vaadin-custom-field` doesn't properly support slotting the inputs in cases like:
```html
<vaadin-custom-field>
  <slot name="date-picker">
    <vaadin-date-picker></vaadin-date-picker>
  </slot>
  <slot name="time-picker">
    <vaadin-time-picker></vaadin-time-picker>
  </slot>
</vaadin-custom-field>
```

The problem is that custom field uses a [`MutationObserver`](https://github.com/vaadin/vaadin-custom-field/blob/master/src/vaadin-custom-field-mixin.html#L122) and [this logic in `__setInputsFromChildNodes()`](https://github.com/vaadin/vaadin-custom-field/blob/master/src/vaadin-custom-field-mixin.html#L147) to populate the `inputs` array (which is used for populating the `value`):
```js
const inputs = Array.from(this.querySelectorAll('*')).filter(node => node.validate || node.checkValidity);
```

In the above problem case this only finds the `vaadin-date-picker` and `vaadin-time-picker` which are the default content for the slots directly in the template (as children of `<slot>`) but it does not detect if those slots get anything slotted to them (which would replace the default content that is rendered).

I think this would be solved by using [`Polymer.FlattenedNodesObserver`](https://polymer-library.polymer-project.org/2.0/docs/devguide/shadow-dom#observe-nodes) on the [`<slot id="slot"></slot>`](https://github.com/vaadin/vaadin-custom-field/blob/master/src/vaadin-custom-field.html#L44) instead of current approach with `MutationObserver` and `querySelectorAll`.

This feature is needed for [`vaadin-date-time-picker`](https://github.com/vaadin/vaadin-date-time-picker).

### PR DESCRIPTION
**Pull request by Haprog**
Fixes #48 including support for wrapping elements.

### CODE REVIEW

File: test/custom-field-slot-test.html [hunk 1]
```diff
@@ -29,6 +29,26 @@
     </script>
   </dom-module>
 
+  <dom-module id="x-field2">
+    <template>
+      <vaadin-custom-field id="customField">
+        <div id="wrapper">
+          <slot></slot>
+        </div>
```
Comment (Haprog) [inline test/custom-field-slot-test.html]: Same as current `<x-field>` but additionally has this wrapper element as a parent for the slot.

File: test/custom-field-slot-test.html [hunk 2]
```diff
@@ -47,6 +67,15 @@
     </template>
   </test-fixture>
 
+  <test-fixture id="nested2">
```
Comment (Haprog) [inline test/custom-field-slot-test.html]: Same as `nested` but uses `x-field2` (with extra wrapper element).

File: test/custom-field-slot-test.html [hunk 3]
```diff
@@ -60,18 +89,31 @@
     </template>
   </test-fixture>
 
+  <test-fixture id="nested-deep2">
```
Comment (Haprog) [inline test/custom-field-slot-test.html]: Same as `nested-deep` but uses `x-field2` (with extra wrapper element).

File: src/vaadin-custom-field-mixin.html [hunk 4]
```diff
@@ -150,19 +150,25 @@
...
+        elements = elem.assignedNodes({flatten: true}).filter(node => node.nodeType === Node.ELEMENT_NODE);
+      } else {
+        result.push(elem);
+        elements = Array.from(elem.children);
+      }
+      elements.forEach(elem => {
+        result.push(...this.__queryAllAssignedElements(elem));
+      });
+      return result;
+    }
```
Comment (Haprog) [inline src/vaadin-custom-field-mixin.html]: Here's an alternate version of the same method that does exactly the same thing (and also passes all tests). I'm just not sure which version I prefer. I think the committed one (above) is more readable (maybe) even though it uses temporary variables (`result` and `elements`).

```js
__queryAllAssignedElements(elem) {
  const flatten = (a, b) => a.concat(b);
  if (elem.tagName === 'SLOT') {
    return elem.assignedNodes({flatten: true})
      .filter(node => node.nodeType === Node.ELEMENT_NODE)
      .map(this.__queryAllAssignedElements.bind(this))
      .reduce(flatten);
  } else {
    return Array.from(elem.children)
      .map(this.__queryAllAssignedElements.bind(this))
      .reduce(flatten, [elem]);
  }
}
```

File: src/vaadin-custom-field-mixin.html [hunk 5]
```diff
@@ -150,19 +150,25 @@
...
+      const result = [];
+      let elements;
+      if (elem.tagName === 'SLOT') {
+        elements = elem.assignedNodes({flatten: true}).filter(node => node.nodeType === Node.ELEMENT_NODE);
+      } else {
+        result.push(elem);
+        elements = Array.from(elem.children);
+      }
+      elements.forEach(elem => {
+        result.push(...this.__queryAllAssignedElements(elem));
```
Comment (yuriy-fix) [inline src/vaadin-custom-field-mixin.html]: Non-blocking nit: is it possible to make one-liner? :)

File: src/vaadin-custom-field-mixin.html (hunk 4 above)
Comment (yuriy-fix) [reply src/vaadin-custom-field-mixin.html]: I would prefer the first option (that is already in the code) as it's more readable.

**Comment by Haprog** [comment]
@yuriy-fix 
> Now with nested inputs support should the component still take care of proper styling (layout) of those. Probably the docs should be updated in order to mention that it's a responsibility of the developer for nested structures

Nested inputs were already supported even before these slotting fixes (previously this was just using `querySelectorAll('*')` and filtering for input elements). This fix only makes it work also for the cases when the user provides a (nested) slot instead of directly providing the (nested) input.

Imo it should be obvious that whatever (possibly complex) DOM the developer provides, also needs to be styled by the developer (except maybe in very simple cases). Docs could be improved I guess but I would do it outside of this PR.

File: src/vaadin-custom-field-mixin.html (hunk 5 above)
Comment (Haprog) [reply src/vaadin-custom-field-mixin.html]: You mean like this just removing few new lines?
```js
elements.forEach(elem => { result.push(...this.__queryAllAssignedElements(elem)); });
```
Or something else?

Comment (yuriy-fix) [reply src/vaadin-custom-field-mixin.html]: Something like:
`elements.forEach(elem => result.push(...this.__queryAllAssignedElements(elem)));`

Comment (Haprog) [reply src/vaadin-custom-field-mixin.html]: Updated. Will merge after tests pass.

**Pull request merged**